core/src/main/java/io/snappydata/impl/SnappyHiveCatalog.java

```diff
@@ -67,13 +71,30 @@
 
   final private static String THREAD_GROUP_NAME = "HiveMetaStore Client Group";
 
+  @SuppressWarnings("UnstableApiUsage")
+  private final com.google.common.cache.Cache<
+      Map.Entry<String, String>, Boolean> nonExistentTables;
+  @SuppressWarnings("UnstableApiUsage")
+  private final com.google.common.cache.LoadingCache<
+      Map.Entry<String, String>, Table> tableCache;
+
   private final Future<?> initFuture;
 
   public static final Object hiveClientSync = new Object();
 
   private final ExecutorService hmsQueriesExecutorService;
 
   public SnappyHiveCatalog() {
+    this.nonExistentTables = CacheBuilder.newBuilder().maximumSize(2000).build();
+    this.tableCache = CacheBuilder.newBuilder().maximumSize(2000).build(
+        new CacheLoader<Map.Entry<String, String>, Table>() {
+          @Override
+          public Table load(@Nonnull Map.Entry<String, String> table) throws Exception {
+            Hive client = Hive.get();
+            return client.getTable(table.getKey(), table.getValue());
```

Comment:
The first lookup will bear the cost and subsequent is fast. Good. But for demo purpose etc we should have run few queries beforehand otherwise the same latency will show up if used first time. Just a note.

Reply:
I didn't see any significant slowdown even in first explicit call. Likely due to internal calls by stats provider etc.